File: old_backend/offerings.js
/**
 * Campaign offerings (products / services) helpers.
 */

export function normalizeOfferings(raw) {
  if (!Array.isArray(raw)) return [];

  return raw
    .map((o) => {
      const keywords = Array.isArray(o.keywords)
        ? o.keywords.map(String).map((s) => s.trim()).filter(Boolean)
        : String(o.keywords || '')
            .split(',')
            .map((s) => s.trim())
            .filter(Boolean);

      return {
        name: String(o.name || '').trim().slice(0, 120),
        type: o.type === 'service' ? 'service' : 'product',
        description: String(o.description || '').trim().slice(0, 500),
        url: String(o.url || '').trim().slice(0, 500),
        price_hint: String(o.price_hint || '').trim().slice(0, 80),
        keywords: keywords.slice(0, 12),
      };
    })
    .filter((o) => o.name.length > 0)
    .slice(0, 30);
}

export function mergeKeywordsFromOfferings(campaignKeywords, offerings) {
  const base = Array.isArray(campaignKeywords) ? [...campaignKeywords] : [];
  const seen = new Set(base.map((k) => k.toLowerCase()));

  for (const o of offerings || []) {
    for (const kw of [o.name, ...(o.keywords || [])]) {
      const t = String(kw || '').trim();
      if (t.length < 2) continue;
      const key = t.toLowerCase();
      if (!seen.has(key)) {
        seen.add(key);
        base.push(t);
      }
    }
  }
  return base.slice(0, 40);
}

File: backend/app/services/offerings.py
def normalize_offerings(raw) -> list[dict]:
    if not isinstance(raw, list):
        return []
    out = []
    for o in raw:
        kws = o.get("keywords") or []
        if isinstance(kws, str):
            kws = [s.strip() for s in kws.split(",") if s.strip()]
        else:
            kws = [str(s).strip() for s in kws if str(s).strip()]
        name = str(o.get("name") or "").strip()[:120]
        if not name:
            continue
        out.append(
            {
                "name": name,
                "type": "service" if o.get("type") == "service" else "product",
                "description": str(o.get("description") or "").strip()[:500],
                "url": str(o.get("url") or "").strip()[:500],
                "price_hint": str(o.get("price_hint") or "").strip()[:80],
                "keywords": kws[:12],
            }
        )
    return out[:30]


def merge_keywords_from_offerings(campaign_keywords: list | None, offerings: list | None) -> list[str]:
    base = list(campaign_keywords or [])
    seen = {k.lower() for k in base}
    for o in offerings or []:
        for kw in [o.get("name"), *(o.get("keywords") or [])]:
            t = str(kw or "").strip()
            if len(t) < 2:
                continue
            key = t.lower()
            if key not in seen:
                seen.add(key)
                base.append(t)
    return base[:40]
